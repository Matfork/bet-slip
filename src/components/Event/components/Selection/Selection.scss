@import "src/assets/scss/media";
@import "src/assets/scss/spacing";
@import "src/assets/scss/colors";

$button-width: 160px;

.Selection-component {
  display: flex;
  flex-direction: column;

  @media only screen and (min-width: $md-min) {
    justify-content: space-around;
    flex-direction: row;

    .Selection-component__button {
      width: $button-width !important;
      margin-bottom: 0 !important;
    }
  }

  .Selection-component__button {
    width: 100%;
    margin-bottom: $spacing-1;

    span {
      display: block;
      text-align: center;
    }
  }

  .MuiButton-root:hover.Mui-disabled,
  .Selection-component__button--selected {
    background-color: $green !important;
    color: $white !important;
  }
}
