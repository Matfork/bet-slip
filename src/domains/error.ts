export interface ErrorResponse {
  status: number;
  message: string;
}

export class ApiError extends Error {
  status: number;

  constructor({ message, status }: { message: string; status: number }) {
    super(message); // (1)
    this.status = status; // (2)
  }
}
