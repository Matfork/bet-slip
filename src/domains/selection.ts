export enum SelectedSelectionOptions {
  Select,
  Unselect,
}

export interface SelectionType {
  id: string;
  name: string;
  price: number;
}
