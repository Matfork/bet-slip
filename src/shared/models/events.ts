export interface EventType {
  id: string;
  name: string;
  markets: MarketType[];
}

interface MarketType {
  id: string;
  name: string;
  markets: SelectionType[];
}

interface SelectionType {
  id: string;
  name: string;
  price: number;
}
