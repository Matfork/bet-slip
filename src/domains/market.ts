import { SelectionType } from './selection';

export interface MarketType {
  id: string;
  name: string;
  selections: SelectionType[];
}
