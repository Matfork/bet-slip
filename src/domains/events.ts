import { MarketType } from './market';

export interface EventType {
  id: string;
  name: string;
  markets: MarketType[];
}
