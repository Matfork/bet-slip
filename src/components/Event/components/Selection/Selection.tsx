import { Button, Typography } from '@material-ui/core';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { SelectionType, SelectedSelectionOptions } from 'domains/selection';
import { EventsActionCreators } from 'store/actions/action-creators/events.actionCreator';
import { RootState } from 'store/reducers';
import { EventsInitialState } from 'store/reducers/events.reducer';
import './Selection.scss';

interface SelectionProps {
  selections: SelectionType[];
}

const className = 'Selection-component';

const SelectionComponent: React.FC<SelectionProps> = ({ selections }: SelectionProps) => {
  const dispatch = useDispatch();
  const { userSelection } = useSelector<RootState, EventsInitialState>((state) => state.events);
  const selectedIds = userSelection.map((sel) => sel.id);

  const handleButtonClicked = (selection: SelectionType): void => {
    dispatch({
      type: EventsActionCreators.SetUserSelection,
      payload: {
        selection,
        option: SelectedSelectionOptions.Select,
      },
    });
  };

  return (
    <div className={className}>
      {selections.map((selection) => {
        const isAlreadySelected = selectedIds.includes(selection.id);
        return (
          <Button
            onClick={() => handleButtonClicked(selection)}
            size="small"
            disableElevation
            variant="outlined"
            key={selection.id}
            className={`${className}__button ${isAlreadySelected ? `${className}__button--selected` : ''}`}
            disabled={isAlreadySelected}
          >
            <Typography variant="body2">{selection.name}</Typography>
            <Typography variant="body2">{selection.price}</Typography>
          </Button>
        );
      })}
    </div>
  );
};

export default SelectionComponent;
