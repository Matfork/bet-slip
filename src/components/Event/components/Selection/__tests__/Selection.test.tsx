import { fireEvent, screen } from '@testing-library/react';
import { customRender } from 'test.utilities';
import { initialStateStub, selectionsStub } from 'test.stub';
import { EventsActionCreators } from 'store/actions/action-creators/events.actionCreator';
import { SelectedSelectionOptions } from 'domains/selection';
import Selection from '../Selection';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  ...(jest.requireActual('react-redux') as {}),
  useDispatch: () => mockDispatch,
}));

describe('Selection Component', () => {
  it('should test snapshot', () => {
    const component = customRender(<Selection selections={selectionsStub} />);
    expect(component.container).toMatchSnapshot();
  });

  it('should display all selections avaiable', () => {
    customRender(<Selection selections={selectionsStub} />);
    const titleEl = screen.queryByText(/Real Madrid/);
    expect(titleEl).toBeInTheDocument();
  });

  it('should not display selections if there is no data', () => {
    customRender(<Selection selections={[]} />);
    const titleEl = screen.queryByText(/Real Madrid/);
    expect(titleEl).toBeNull();
  });

  it('should dispatch action on button clicked', () => {
    customRender(<Selection selections={selectionsStub} />);

    const button = screen.queryByText(/Real Madrid/)?.closest('button')!;
    fireEvent.click(button);

    expect(mockDispatch).toHaveBeenCalledWith({
      type: EventsActionCreators.SetUserSelection,
      payload: {
        selection: selectionsStub[0],
        option: SelectedSelectionOptions.Select,
      },
    });
  });

  it('should display already selected options with custom selected class', () => {
    customRender(<Selection selections={selectionsStub} />, {
      initialState: {
        ...initialStateStub,
        events: { ...initialStateStub.events, userSelection: [selectionsStub[0]] },
      },
    });

    const button = screen.queryByText(/Real Madrid/)?.closest('button')!;
    expect(button.classList.contains('Selection-component__button--selected')).toBeTruthy();
  });
});
