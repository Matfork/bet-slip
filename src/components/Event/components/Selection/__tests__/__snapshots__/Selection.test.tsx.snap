// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selection Component should test snapshot 1`] = `
<div>
  <div
    class="Selection-component"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined Selection-component__button  MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Real Madrid
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          1.23
        </p>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined Selection-component__button  MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Barcelona
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          2.05
        </p>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
